python/lsst/ts/observatory/control/auxtel/atcs.py

```diff
@@ -1524,6 +1524,140 @@ def unset_azel_slew_checks(self, checks):
         self.check.adome = checks.dome
         self.check.atdometrajectory = checks.dometrajectory
 
+    async def _ready_to_take_data(self):
+        """Wait until ATCS is ready to take data.
+
+        If cancelled or some condition times out or raise an exeception it will
+        set the result of the future to the exception so it propagates to
+        whomever is waiting on it.
+        """
+        # Things to check
+        # 1 - ATMCS evt_allAxesInPosition.inPosition == True
+        # 2 - All ATAOS corrections are applied
+        # 3 - ATMCS and ATDome positions are within specified range.
+        ready = False
+        try:
+            while not ready:
+                # This loop will run until all conditions are met. Note that it
+                # may happen that one condition will be me when we run it the
+                # first time and then not when we run the second time.
+                check = await asyncio.gather(
+                    self.atmcs_in_position(),
+                    self.ataos_corrections_completed(),
+                    self.dome_az_in_position.wait(),
+                )
+                self.log.debug(
+                    f"Ready to take data:: atmcs={check[0]}, ataos={check[1]}, atdome={check[2]}."
+                )
+                ready = all(check)
+        except asyncio.CancelledError as e:
+            self._ready_to_take_data_future.set_exception(e)
+        except asyncio.TimeoutError as e:
+            self._ready_to_take_data_future.set_exception(e)
+        except Exception as e:
+            self._ready_to_take_data_future.set_exception(e)
```

Comment:
Is it necessary to split out these different errors like this? In the end they all call the same method so lines 1553 to 1556 are superfluous.

Reply:
yes, the problem is that some exceptions are no longer children of `Exception`, like `CancelledError`. So I am future-proofing the code if they ever decide to make the same with `TimeoutError`.

Reply:
OK, good idea.